File: app-ts/src/BaseClasses/EventDrivenCode/Delegator.ts
export class EventDelegator<T = void> {
    private _methods: Record<string, (value: T) => void> = {};

    constructor() {}

    public addMethod(event: (value: T) => void, key: string): void {
        if (this._methods[key]) {
            console.error(`Key already exists: ${key}`);
            return;
        }
        this._methods[key] = event;
    }

    public invoke(value: T, key: string | null = null): void {
        if (key === null) {
            for (let key in this._methods) {
                this._methods[key](value);
            }
        } else {
            const method = this._methods[key];
            if (method) {
                method(value);
            } else {
                console.error(`Method not found for key: ${key}`);
            }
        }
    }

    public invokeByQueue(value: T, keys: string[]): void {
        for (let key of keys) {
            const method = this._methods[key];
            if (method) {
                method(value);
            } else {
                console.error(`Method not found for key: ${key}`);
            }
        }
    }

    public removeMethod(key: string): void {
        delete this._methods[key];
    }

    public clearMethods(): void {
        this._methods = {};
    }
}

export class ActionDelegator<T = void, R = void> {
    private _method: ((value: T) => R) | null = null;

    constructor() {}

    public setMethod(event: (value: T) => R): void {
        this._method = event;
    }

    public invoke(value: T): R | void {
        if (this._method) {
            return this._method(value);
        } else {
            console.error(`Method not set`);
        }
    }

    public clearMethod(): void {
        this._method = null;
    }
}

File: api/Extend/RxTool/Delegator.py
import logging
from typing import Dict, Callable, TypeVar, Generic, List, Optional, Any, Union, AsyncIterable
import asyncio

from api.Extend.ExtendFunc import ExtendFunc

logger = logging.getLogger(__name__)

T = TypeVar('T')
R = TypeVar('R')

class EventDelegator(Generic[T]):
    def __init__(self):
        self._methods: Dict[str, Callable[[T], None]] = {}
    
    def add_method(self, event: Callable[[T], None], key: str) -> None:
        if key in self._methods:
            return
        self._methods[key] = event
    
    def has_method(self, key: str) -> bool:
        return key in self._methods
    
    def invoke(self, value: T, key: Optional[str] = None) -> None:
        if key is None:
            # コールバック実行中の変更を防ぐためコピー
            methods_copy = dict(self._methods)
            for k, method in methods_copy.items():
                try:
                    method(value)
                except Exception as e:
                    ExtendFunc.ExtendPrint(f"Error invoking method {k}: {e}")
        else:
            method = self._methods.get(key)
            if method:
                method(value)
            else:
                ExtendFunc.ExtendPrint(f"Method not found for key: {key}")
    
    def invoke_by_queue(self, value: T, keys: List[str]) -> None:
        for key in keys:
            method = self._methods.get(key)
            if method:
                method(value)
            else:
                ExtendFunc.ExtendPrint(f"Method not found for key: {key}")
    
    def remove_method(self, key: str) -> None:
        if key in self._methods:
            del self._methods[key]
    
    def clear_methods(self) -> None:
        self._methods.clear()


class ActionDelegator(Generic[T, R]):
    def __init__(self):
        self._method: Optional[Callable[[T], R]] = None
    
    def set_method(self, event: Callable[[T], R]) -> None:
        self._method = event
    
    def invoke(self, value: T) -> Optional[R]:
        if self._method:
            return self._method(value)
        else:
            ExtendFunc.ExtendPrint("Method not set")
            return None
    
    def clear_method(self) -> None:
        self._method = None